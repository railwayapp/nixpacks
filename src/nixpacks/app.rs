use path_slash::PathBufExt;
use std::collections::BTreeMap;
use std::ffi::OsStr;
use std::path::Path;
use std::{env, fs, path::PathBuf};

use anyhow::{bail, Context, Result};
use globset::Glob;
use ignore::{DirEntry, WalkBuilder};
use regex::Regex;
use serde::de::DeserializeOwned;

pub type StaticAssets = BTreeMap<String, String>;

pub const ASSETS_DIR: &str = "/assets/";

#[derive(Debug, Clone)]
pub struct App {
    pub source: PathBuf,
    pub paths: Vec<PathBuf>,
}

impl App {
    pub fn new(path: &str) -> Result<App> {
        let current_dir = env::current_dir()?;
        let source = current_dir
            .join(path)
            .canonicalize()
            .context("Failed to read app source directory")?;

        let dir = fs::read_dir(source.clone()).context("Failed to read app source directory")?;
        let paths: Vec<PathBuf> = dir.map(|path| path.unwrap().path()).collect();

        Ok(App { source, paths })
    }

    /// Check if a file exists
    pub fn includes_file(&self, name: &str) -> bool {
        self.source.join(name).is_file()
    }

    /// Returns a list of paths matching a glob pattern
    ///
    /// # Errors
    /// Creating the Glob fails
    pub fn find_files(&self, pattern: &str) -> Result<Vec<PathBuf>> {
        let directories = self
            .find_glob(pattern)?
            .into_iter()
            .filter(|path| path.is_file())
            .collect();

        Ok(directories)
    }

    /// Returns a list of paths matching a glob pattern
    ///
    /// # Errors
    /// Creating the Glob fails
    pub fn find_directories(&self, pattern: &str) -> Result<Vec<PathBuf>> {
        let directories = self
            .find_glob(pattern)?
            .into_iter()
            .filter(|path| path.is_dir())
            .collect();

        Ok(directories)
    }

    fn find_glob(&self, pattern: &str) -> Result<Vec<PathBuf>> {
        let full_pattern = self.source.join(pattern);

        let pattern_str = match full_pattern.to_str() {
            Some(s) => s,
            None => return Ok(Vec::new()),
        };

        let walker = WalkBuilder::new(&self.source)
            // this includes hidden directories & files
            .hidden(false)
            .sort_by_file_name(OsStr::cmp)
            .build();
        let glob = Glob::new(pattern_str)?.compile_matcher();

        let relative_paths = walker
            .into_iter()
            .filter_map(Result::ok) // remove bad ones
            .map(DirEntry::into_path) // convert to paths
            .filter(|path| glob.is_match(path)) // find matches
            .collect();

        Ok(relative_paths)
    }

    /// Check if a path matching a glob exists
    pub fn has_match(&self, pattern: &str) -> bool {
        match self.find_files(pattern) {
            Ok(v) => !v.is_empty(),
            Err(_e) => false,
        }
    }

    /// Read the contents of a file
    ///
    /// # Errors
    /// This will error if the path doesn't exist, or if the contents isn't UTF-8
    pub fn read_file(&self, name: &str) -> Result<String> {
        let data = fs::read_to_string(PathBuf::from_slash_lossy(
            self.source.join(name).as_os_str(),
        ))
        .with_context(|| {
            let relative_path = self.strip_source_path(Path::new(name)).unwrap();
            format!("Error reading {}", relative_path.to_str().unwrap())
        })?;

        Ok(data.replace("\r\n", "\n"))
    }

    pub fn find_match(&self, re: &Regex, pattern: &str) -> Result<bool> {
        let paths = match self.find_files(pattern) {
            Ok(v) => v,
            Err(_e) => return Ok(false),
        };

        for path in paths {
            let path_buf = fs::canonicalize(path)?;

            if let Some(p) = path_buf.to_str() {
                let f = self.read_file(p)?;
                if re.find(f.as_str()).is_some() {
                    return Ok(true);
                }
            }
        }

        Ok(false)
    }

    /// Check if a directory exists
    pub fn includes_directory(&self, name: &str) -> bool {
        self.source.join(name).is_dir()
    }

    #[cfg(target_os = "windows")]
    pub fn is_file_executable(&self, name: &str) -> bool {
        true
    }

    /// Check if a path is an executable file
    #[cfg(not(target_os = "windows"))]
    pub fn is_file_executable(&self, name: &str) -> bool {
        use std::os::unix::prelude::PermissionsExt;

        let path = self.source.join(name);
        if path.is_file() {
            let metadata = path.metadata().unwrap();
            metadata.permissions().mode() & 0o111 != 0
        } else {
            false
        }
    }

    pub fn read_json<T>(&self, name: &str) -> Result<T>
    where
        T: DeserializeOwned,
    {
        let contents = self.read_file(name)?;
        let value: T = serde_json::from_str(contents.as_str()).with_context(|| {
            let relative_path = self.strip_source_path(Path::new(name)).unwrap();
            format!("Error reading {} as JSON", relative_path.to_str().unwrap())
        })?;
        Ok(value)
    }

    pub fn read_toml<T>(&self, name: &str) -> Result<T>
    where
        T: DeserializeOwned,
    {
        let contents = self.read_file(name)?;
        let toml_file = toml::from_str(contents.as_str()).with_context(|| {
            let relative_path = self.strip_source_path(Path::new(name)).unwrap();
            format!("Error reading {} as TOML", relative_path.to_str().unwrap())
        })?;
        Ok(toml_file)
    }

    pub fn read_yaml<T>(&self, name: &str) -> Result<T>
    where
        T: DeserializeOwned,
    {
        let contents = self.read_file(name)?;
        let yaml_file = serde_yaml::from_str(contents.as_str())?;
        Ok(yaml_file)
    }

    /// Convert an absolute path to a path relative to the app source directory
    pub fn strip_source_path(&self, abs_path: &Path) -> Result<PathBuf> {
        let source_str = match self.source.to_str() {
            Some(s) => s,
            None => bail!("Failed to parse source path"),
        };

        // Strip source path from absolute path
        let stripped = match abs_path.strip_prefix(source_str) {
            Ok(p) => p,
            Err(_e) => abs_path,
        };

        // Convert path to PathBuf
        Ok(stripped.to_owned())
    }

    /// Get the path in the container to an asset defined in `static_assets`.
    pub fn asset_path(&self, name: &str) -> String {
        format!("{}{}", ASSETS_DIR, name)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use serde::{Deserialize, Serialize};
    use serde_json::{Map, Value};
    use std::collections::HashMap;

    #[derive(Serialize, Deserialize)]
    struct TestPackageJson {
        name: String,
        scripts: HashMap<String, String>,
    }

    #[test]
    fn test_creates_app() -> Result<()> {
        let app = App::new("./examples/node-npm")?;
        assert_eq!(app.paths.len(), 5);
        Ok(())
    }

    #[test]
    fn test_read_file() -> Result<()> {
        let app = App::new("./examples/node-npm")?;
        assert_eq!(
            app.read_file("index.ts")?.trim_end(),
            "console.log(\"Hello from NPM\");"
        );
        Ok(())
    }

    #[test]
    #[cfg(not(target_os = "windows"))]
    fn test_is_file_executable() -> Result<()> {
        let app = App::new("./examples/java-gradle-hello-world")?;
        assert!(app.is_file_executable("gradlew"));
        assert!(!app.is_file_executable("build.gradle"));
        Ok(())
    }

    #[test]
    fn test_read_json_file() -> Result<()> {
        let app = App::new("./examples/node-npm")?;
        let value: Map<String, Value> = app.read_json("package.json")?;
        assert!(value.get("name").is_some());
        assert_eq!(value.get("name").unwrap(), "npm");
        Ok(())
    }

    #[test]
    fn test_read_structured_json_file() -> Result<()> {
        let app = App::new("./examples/node-npm")?;
        let value: TestPackageJson = app.read_json("package.json")?;
        assert_eq!(value.name, "npm");
        assert_eq!(value.scripts.get("build").unwrap(), "tsc -p tsconfig.json");
        Ok(())
    }

    #[test]
    fn test_read_toml_file() -> Result<()> {
        let app = App::new("./examples/rust-rocket")?;
        let toml_file: toml::Value = app.read_toml("Cargo.toml")?;
        assert!(toml_file.get("package").is_some());
        assert_eq!(
            toml_file
                .get("package")
                .unwrap()
                .get("name")
                .unwrap()
                .as_str()
                .unwrap(),
            "rocket"
        );
        Ok(())
    }

    #[test]
    fn test_find_files() -> Result<()> {
        let app = App::new("./examples/node-monorepo")?;
        let m = app.find_files("**/*.tsx").unwrap();
        let dir = env::current_dir().unwrap();
        assert_eq!(
            m,
            vec![
                dir.join("examples/node-monorepo/packages/client/pages/_app.tsx")
                    .canonicalize()?,
                dir.join("examples/node-monorepo/packages/client/pages/index.tsx")
                    .canonicalize()?
            ]
        );
        Ok(())
    }

    #[test]
    fn test_find_match() -> Result<()> {
        let app = App::new("./examples/node-monorepo")?;
        let re = Regex::new(r"className")?;
        let m = app.find_match(&re, "**/*.tsx").unwrap();
        assert!(m);
        Ok(())
    }

    #[test]
    fn test_strip_source_path() -> Result<()> {
        let app = App::new("./examples/node-npm")?;
        let path_to_strip = app.source.join("foo/bar.txt");
        assert_eq!(
            &app.strip_source_path(&path_to_strip).unwrap(),
            Path::new("foo/bar.txt")
        );
        Ok(())
    }

    #[test]
    fn test_strip_source_path_no_source_prefix() -> Result<()> {
        let app = App::new("./examples/node-npm")?;
        assert_eq!(
            &app.strip_source_path(Path::new("no/prefix.txt"))?,
            Path::new("no/prefix.txt")
        );
        Ok(())
    }

    #[test]
    fn test_static_asset_path() -> Result<()> {
        let app = App::new("./examples/node-npm")?;
        assert_eq!(&app.asset_path("hi.txt"), "/assets/hi.txt");
        Ok(())
    }
}
