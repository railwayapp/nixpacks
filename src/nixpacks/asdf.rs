use regex::Regex;
use std::collections::HashMap;

pub fn parse_tool_versions_content(file_content: &str) -> HashMap<String, String> {
    let re = Regex::new(r"\s+").unwrap();
    file_content
        .lines()
        .map(str::trim)
        .filter(|line| !line.starts_with('#'))
        .filter_map(|line| {
            let parts: Vec<&str> = re.splitn(line, 2).collect();
            if parts.len() == 2 {
                Some((parts[0].to_string(), parts[1].to_string()))
            } else {
                None
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_tool_versions_content() {
        let file_content = "\
# This is a comment
python 3.10.4
 poetry  1.7.1
# exact ruby version required for homebrew
ruby\t3.1.4
direnv  2.32.3 ";
        let versions = parse_tool_versions_content(file_content);

        let expected = HashMap::from([
            ("direnv".to_string(), "2.32.3".to_string()),
            ("python".to_string(), "3.10.4".to_string()),
            ("ruby".to_string(), "3.1.4".to_string()),
            ("poetry".to_string(), "1.7.1".to_string()),
        ]);

        assert_eq!(versions, expected);
    }
}
