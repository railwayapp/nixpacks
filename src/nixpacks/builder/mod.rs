use super::{environment::Environment, plan::BuildPlan};
use anyhow::Result;
use async_trait::async_trait;

pub mod docker;

#[async_trait]
pub trait ImageBuilder {
    async fn create_image(
        &self,
        app_source: &str,
        plan: &BuildPlan,
        env: &Environment,
    ) -> Result<()>;
}
