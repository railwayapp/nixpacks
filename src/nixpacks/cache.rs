pub fn sanitize_cache_key(cache_key: String) -> String {
    cache_key
        .chars()
        .filter(|x| !matches!(x, '.')) // remove dot from the string
        .map(|x| match x {
            ' ' => '-',
            _ => x,
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sanitizing_cache_key() {
        assert_eq!(sanitize_cache_key("key".to_string()), "key".to_string());
        assert_eq!(
            sanitize_cache_key("s p a c e s".to_string()),
            "s-p-a-c-e-s".to_string()
        );
        assert_eq!(
            sanitize_cache_key("s/my-cache-key".to_string()),
            "s/my-cache-key".to_string()
        );
        assert_eq!(sanitize_cache_key("/.m2".to_string()), "/m2".to_string());
    }
}
