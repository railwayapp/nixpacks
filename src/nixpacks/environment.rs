use anyhow::Result;
use regex::Regex;
use std::{collections::BTreeMap, env};

pub type EnvironmentVariables = BTreeMap<String, String>;

#[derive(Default, Debug)]
pub struct Environment {
    variables: EnvironmentVariables,
}

impl Environment {
    pub fn new(variables: EnvironmentVariables) -> Environment {
        Environment { variables }
    }

    pub fn from_envs(envs: Vec<&str>) -> Result<Environment> {
        let mut environment = Environment::default();
        for env in envs {
            let matches = Regex::new(r"([A-Za-z0-9_-]*)(?:=?)([\s\S]*)")
                .unwrap()
                .captures(env)
                .unwrap();
            if matches.get(2).unwrap().as_str() == "" {
                // No value, pull from the current environment
                let name = matches.get(1).unwrap().as_str();
                if let Ok(value) = env::var(name) {
                    environment.set_variable(name.to_string(), value);
                }
            } else {
                // Use provided name, value pair
                environment.set_variable(
                    matches.get(1).unwrap().as_str().to_string(),
                    matches.get(2).unwrap().as_str().to_string(),
                );
            }
        }

        Ok(environment)
    }

    pub fn get_variable(&self, name: &str) -> Option<&str> {
        self.variables.get(name).map(String::as_str)
    }

    pub fn get_config_variable(&self, name: &str) -> Option<String> {
        self.get_variable(format!("NIXPACKS_{}", name).as_str())
            .map(|var| var.replace('\n', ""))
    }

    pub fn is_config_variable_truthy(&self, name: &str) -> bool {
        if let Some(var) = self.get_config_variable(name) {
            matches!(var.as_str(), "1" | "true")
        } else {
            false
        }
    }

    pub fn set_variable(&mut self, name: String, value: String) {
        self.variables.insert(name, value);
    }

    pub fn get_variable_names(&self) -> Vec<String> {
        self.variables.keys().cloned().collect()
    }

    pub fn clone_variables(env: &Environment) -> EnvironmentVariables {
        env.variables.clone()
    }
}

#[cfg(test)]
mod tests {
    use super::Environment;

    #[test]
    fn set_and_get_variables() {
        let mut environment = Environment::default();
        assert!(environment.get_variable("hello").is_none());
        environment.set_variable("hello".to_string(), "world".to_string());
        assert_eq!(environment.get_variable("hello"), Some("world"));
    }

    #[test]
    fn test_environment_variable_parsing() {
        let environment =
            Environment::from_envs(vec!["HELLO=world", "CARGO_PKG_NAME", "NON_EXISTANT"]).unwrap();
        assert_eq!(environment.get_variable("HELLO"), Some("world"));
        assert_eq!(environment.get_variable("CARGO_PKG_NAME"), Some("nixpacks"));
        assert!(environment.get_variable("NON_EXISTANT").is_none());
    }

    #[test]
    fn test_create_equals_sign_parsing() {
        let environment = Environment::from_envs(vec!["INVALID=ENV=CONFIG"]).unwrap();
        assert_eq!(environment.get_variable("INVALID"), Some("ENV=CONFIG"));
    }

    #[test]
    fn test_get_config_variable() {
        let mut environment = Environment::default();
        environment.set_variable("NIXPACKS_HELLO".to_string(), "world".to_string());
        assert_eq!(
            environment.get_config_variable("HELLO"),
            Some("world".to_string())
        );
    }

    #[test]
    fn test_get_config_variable_truthy() {
        let mut environment = Environment::default();

        environment.set_variable("NIXPACKS_YES".to_string(), "1".to_string());
        environment.set_variable("NIXPACKS_NO".to_string(), "0".to_string());

        assert!(environment.is_config_variable_truthy("YES"));
        assert!(!environment.is_config_variable_truthy("NO"));
    }

    #[test]
    fn test_get_config_variable_strips_newlines() {
        let mut environment = Environment::default();
        environment.set_variable("NIXPACKS_BUILD_CMD".to_string(), "hello\nworld".to_string());
        assert_eq!(
            environment.get_config_variable("BUILD_CMD"),
            Some("helloworld".to_string())
        );
    }
}
