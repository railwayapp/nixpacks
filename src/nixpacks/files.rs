use anyhow::Result;
use ignore::WalkBuilder;
use std::{fs, io, path::Path};

pub fn recursive_copy_dir<T: AsRef<Path>, Q: AsRef<Path>>(source: T, dest: Q) -> Result<()> {
    let walker = WalkBuilder::new(&source)
        .follow_links(false)
        // this includes hidden directories & files
        .standard_filters(false)
        .hidden(false)
        .build();

    for entry in walker {
        let entry = entry?;

        if let Some(file_type) = entry.file_type() {
            let from = entry.path();
            let to = dest.as_ref().join(from.strip_prefix(&source)?);

            // create directories
            if file_type.is_dir() {
                if let Err(e) = fs::create_dir(to) {
                    match e.kind() {
                        io::ErrorKind::AlreadyExists => {}
                        _ => return Err(e.into()),
                    }
                }
            }
            // copy files
            else if file_type.is_file() {
                fs::copy(from, &to)?;
                // replace CRLF with LF
                if let Ok(data) = fs::read_to_string(from) {
                    fs::write(&to, data.replace("\r\n", "\n"))?;
                }
            }
        }
    }
    Ok(())
}
