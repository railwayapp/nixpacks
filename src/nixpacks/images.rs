pub const DEFAULT_BASE_IMAGE: &str = "ghcr.io/railwayapp/nixpacks:debian-1669680335";
pub const DEBIAN_SLIM_IMAGE: &str = "debian:bullseye-slim";
