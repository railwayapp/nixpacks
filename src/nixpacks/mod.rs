pub mod app;
pub mod builder;
pub mod environment;
mod files;
pub mod images;
pub mod logger;
pub mod nix;
pub mod phase;
pub mod plan;

pub const NIX_PACKS_VERSION: &str = env!("CARGO_PKG_VERSION");
