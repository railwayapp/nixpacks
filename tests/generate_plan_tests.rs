use nixpacks::{generate_build_plan, nixpacks::plan::generator::GeneratePlanOptions};
use std::env::consts::ARCH;

test_helper::generate_plan_tests!();

#[test]
fn test_custom_plan_path() {
    let plan = generate_build_plan(
        "./examples/custom-plan-path",
        Vec::new(),
        &GeneratePlanOptions {
            config_file: Some("custom-nixpacks.toml".to_string()),
            ..Default::default()
        },
    )
    .unwrap();

    assert_plan_snapshot!(plan);
}

#[test]
fn test_custom_rust_version() {
    let plan = simple_gen_plan("./examples/rust-custom-version");
    let setup = plan.get_phase("setup").unwrap().clone();
    let build = plan.get_phase("build").unwrap().clone();

    assert_eq!(
        build.cmds,
        Some(vec![
            format!("mkdir -p bin"),
            format!("cargo build --release --target {}-unknown-linux-musl", ARCH),
            format!(
                "cp target/{}-unknown-linux-musl/release/rust-custom-version bin",
                ARCH
            )
        ])
    );
    assert_eq!(
        setup
            .nix_pkgs
            .unwrap()
            .iter()
            .filter(|p| p.contains("1.56.0"))
            .count(),
        1
    );
}

#[test]
fn test_rust_rocket() {
    let plan = simple_gen_plan("./examples/rust-rocket");
    let build = plan.get_phase("build").unwrap();
    let start = plan.start_phase.clone().unwrap();

    assert_eq!(
        build.cmds,
        Some(vec![
            format!("mkdir -p bin"),
            format!("cargo build --release --target {}-unknown-linux-musl", ARCH),
            format!("cp target/{}-unknown-linux-musl/release/rocket bin", ARCH)
        ])
    );
    assert!(start.cmd.is_some());
    assert_eq!(start.clone().cmd.unwrap(), "./rocket".to_string());
    assert!(start.run_image.is_some());
}

#[test]
fn test_rust_rocket_no_musl() {
    let plan = generate_build_plan(
        "./examples/rust-rocket",
        vec!["NIXPACKS_NO_MUSL=1"],
        &GeneratePlanOptions::default(),
    )
    .unwrap();
    assert_plan_snapshot!(plan);
}

#[test]
fn test_rust_cargo_workspaces() {
    let plan = simple_gen_plan("./examples/rust-cargo-workspaces");
    let build = plan.get_phase("build").unwrap();

    assert_eq!(
        build.clone().cmds.unwrap()[1],
        format!(
            "cargo build --release --package binary --target {}-unknown-linux-musl",
            ARCH
        )
    );
}

#[test]
fn test_haskell_stack() {
    let plan = simple_gen_plan("./examples/haskell-stack");
    let install = plan.get_phase("install").unwrap();
    let build = plan.get_phase("build").unwrap();
    let start = plan.start_phase.clone().unwrap();

    assert_eq!(install.cmds, Some(vec!["stack setup".to_string()]));
    assert_eq!(build.cmds, Some(vec!["stack install".to_string()]));
    assert_eq!(
        start.cmd,
        Some("/root/.local/bin/haskell-stack-exe".to_string())
    );
}

#[cfg(any(target_arch = "aarch64", target_arch = "x86_64", target_arch = "i386"))]
#[test]
fn test_zig_gyro() {
    let plan = simple_gen_plan("./examples/zig-gyro");
    let install = plan.get_phase("install").unwrap().clone();
    let build = plan.get_phase("build").unwrap();
    let start = plan.start_phase.clone().unwrap();

    assert_eq!(
        build.cmds,
        Some(vec!["zig build -Drelease-safe=true".to_string()])
    );
    assert_eq!(start.cmd, Some("./zig-out/bin/zig-gyro".to_string()));
    assert!(install
        .cmds
        .unwrap()
        .get(0)
        .unwrap()
        .contains("mkdir /gyro"));
}

#[test]
fn test_node_turborepo_custom_app() {
    let plan = generate_build_plan(
        "./examples/node-turborepo",
        vec!["NIXPACKS_TURBO_APP_NAME=docs"],
        &GeneratePlanOptions::default(),
    )
    .unwrap();
    assert!(plan.start_phase.unwrap().cmd.unwrap().contains("docs"));
}
