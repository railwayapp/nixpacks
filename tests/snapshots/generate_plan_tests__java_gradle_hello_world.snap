---
source: tests/generate_plan_tests.rs
expression: plan
---
{
  "providers": [],
  "buildImage": "[build_image]",
  "variables": {
    "NIXPACKS_METADATA": "java"
  },
  "phases": {
    "build": {
      "name": "build",
      "dependsOn": [
        "setup"
      ],
      "cmds": [
        "./gradlew build -x check"
      ],
      "cacheDirectories": [
        "/root/.gradle"
      ]
    },
    "setup": {
      "name": "setup",
      "nixPkgs": [
        "jdk",
        "gradle"
      ],
      "nixOverlays": [],
      "nixpkgsArchive": "[archive]"
    }
  },
  "start": {
    "cmd": "java $JAVA_OPTS -jar  build/libs/*.jar"
  }
}
